### PR TITLE
Rework handling of compositions and projects, and composer windows

### PR DESCRIPTION
**Pull request by nyalldawson**
This PR implements a major rework to how compositions are managed. 

Highlights are:

- A new QgsLayoutManager class which is used for storage and serializing/deserializing of compositions. A QgsLayoutManager is attached to QgsProject. This allows core code to access the compositions attached to a project. All handling of compositions has been moved from app to core, making it easy for server (and other non-app based projects. QField/roam?) to access project compositions without resorting to fragile xml parsing.
- Composer windows are created on demand and destroyed on close. This avoids the heavy requirement of creating all composer windows and widgets for every single composer item when loading projects. It's the main reason loading projects with compositions is so slow currently. Now, composer windows are created only when the window is opened and destroyed when the window is closed. Composer item configuration widgets are also only created on demand (when an item is selected) instead of precreating all widgets.
- A side benefit of destroying composer windows on close is that we now no longer suffer from a Qt bug which blocked use of floating docks in composer windows. Now you can float the item/composition/etc panels! Much nicer for full screen working with compositions on multi-monitor setups.
- Reworked API for working with composers through iface. Since composer windows don't exist unless they are currently open, I've redone all the iface methods for interacting with composers to make it clear that they only apply to open windows (code which works with compositions (not composers) should be updated to use the layoutManager() attached to the active project instead, since compositions in the manager always exist). Additionally, I've added a simple composer interface which provides a stable API for plugins and scripts to work with composer windows. Very basic for now, but in future we can extend with more hooks to allow plugins to interact with composer windows.
- Lots of code cleanups and removal

Sponsored by the Swiss QGIS User Group!

### CODE REVIEW

File: python/core/composer/qgslayoutmanager.sip [hunk 1]
```diff
@@ -0,0 +1,34 @@
+
+class QgsLayoutManager : QObject
+{
+%TypeHeaderCode
+#include <qgslayoutmanager.h>
+%End
+  public:
+
+    explicit QgsLayoutManager( QgsProject* project = 0 );
```
Comment (wonder-sk) [inline python/core/composer/qgslayoutmanager.sip]: Probably /TransferThis/ annotation should be here?

File: python/core/composer/qgslayoutmanager.sip [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    bool addComposition( QgsComposition* composition /Transfer/ );
+    bool removeComposition( QgsComposition* composition );
+
+    void clear();
+    QList< QgsComposition * > compositions() const;
+    QgsComposition *compositionByName( const QString &name ) const;
+    bool readXml( const QDomElement &element, const QDomDocument &doc );
+    QDomElement writeXml( QDomDocument &doc ) const;
+    bool saveAsTemplate( const QString &name, QDomDocument &doc ) const;
+    QgsComposition *duplicateComposition( const QString &name, const QString &newName );
```
Comment (wonder-sk) [inline python/core/composer/qgslayoutmanager.sip]: /Factory/ annotation for duplicateComposition()?

**Review by wonder-sk — COMMENTED** [review]
This looks great! I haven't looked in detail at all the changes but the architecture changes are exactly what we needed for a long time...

**Comment by m-kuhn** [comment]
Nice one!

I wouldn't mind having a new composer module next to core/gui/analysis, after all for custom android apps it will probably not always be required and then the app size could be reduced by stripping the library out, but I'm not sure how much it's interconnected with other core functionality. Most important is: it's out of gui.

**Comment by timlinux** [comment]
Nice stuff @nyalldawson ! I added your nice commit message to a place holder entry in changelog.

http://changelog.qgis.org/en/qgis/version/3.0.0/#map-composer-overhaul

Are there any gotchas plugin / python users will have to look out for if they are constructing composer templates programmatically and need to port their code?

**Comment by nyalldawson** [comment]
> Are there any gotchas plugin / python users will have to look out for if they are constructing composer templates programmatically and need to port their code?

It depends. It's all documented in the api break dox, but if you're just using creating and working with compositions outside of projects, and not using the iface methods, then nothing's changed (yet). It's only if you're interacting with open composers within the iface.

File: python/core/composer/qgslayoutmanager.sip (hunk 1 above)
Comment (nyalldawson) [inline python/core/composer/qgslayoutmanager.sip]: Fixed

File: python/core/composer/qgslayoutmanager.sip (hunk 2 above)
Comment (nyalldawson) [inline python/core/composer/qgslayoutmanager.sip]: No, it's not required - the new composition is automatically added to the manager and ownership remains with the manager.

**Comment by nirvn** [comment]
@nyalldawson , working flawlessly on my machine. Project loading is now blazingly fast, wouhou!

**Comment by NathanW2** [comment]
Nice work!

**Pull request merged**